diff -upNr xf86-video-intel-2.12.0.orign/src/i830_driver.c xf86-video-intel-2.12.0/src/i830_driver.c
--- xf86-video-intel-2.12.0.orign/src/i830_driver.c	2010-09-13 20:59:08.722629729 +0200
+++ xf86-video-intel-2.12.0/src/i830_driver.c	2010-09-13 20:59:46.783566903 +0200
@@ -865,16 +865,8 @@ I830BlockHandler(int i, pointer blockDat
 	intel->BlockHandler = screen->BlockHandler;
 	screen->BlockHandler = I830BlockHandler;
 
-	if (scrn->vtSema) {
-		/* Emit a flush of the rendering cache, or on the 965 and beyond
-		 * rendering results may not hit the framebuffer until significantly
-		 * later.
-		 */
-		intel_batch_submit(scrn,
-				   intel->need_mi_flush ||
-				   !list_is_empty(&intel->flush_pixmaps));
+	if (scrn->vtSema == TRUE)
 		drmCommandNone(intel->drmSubFD, DRM_I915_GEM_THROTTLE);
-	}
 
 	i830_uxa_block_handler(screen);
 
@@ -1072,6 +1064,24 @@ I830UeventFini(ScrnInfoPtr scrn)
     }
 }
 
+static void
+intel_flush_callback(CallbackListPtr *list,
+		     pointer user_data, pointer call_data)
+{
+	ScrnInfoPtr scrn = user_data;
+	intel_screen_private *intel = intel_get_screen_private(scrn);
+
+	if (scrn->vtSema) {
+		/* Emit a flush of the rendering cache, or on the 965
+		 * and beyond rendering results may not hit the
+		 * framebuffer until significantly later.
+		 */
+		intel_batch_submit(scrn,
+				   intel->need_mi_flush ||
+				   !list_is_empty(&intel->flush_pixmaps));
+	}
+}
+
 static Bool
 I830ScreenInit(int scrnIndex, ScreenPtr screen, int argc, char **argv)
 {
@@ -1248,6 +1258,9 @@ I830ScreenInit(int scrnIndex, ScreenPtr 
 	intel->BlockHandler = screen->BlockHandler;
 	screen->BlockHandler = I830BlockHandler;
 
+	if (!AddCallback(&FlushCallback, intel_flush_callback, scrn))
+		return FALSE;
+
 	screen->SaveScreen = xf86SaveScreen;
 	intel->CloseScreen = screen->CloseScreen;
 	screen->CloseScreen = I830CloseScreen;
@@ -1402,6 +1415,8 @@ static Bool I830CloseScreen(int scrnInde
 		intel->front_buffer = NULL;
 	}
 
+	DeleteCallback(&FlushCallback, intel_flush_callback, scrn);
+
 	intel_batch_teardown(scrn);
 
 	if (IS_I965G(intel))

