Index: xserver-xorg-video-i810-modesetting/src/i830_modes.c
===================================================================
--- xserver-xorg-video-i810-modesetting.orig/src/i830_modes.c	2006-10-15 01:03:35.000000000 +1000
+++ xserver-xorg-video-i810-modesetting/src/i830_modes.c	2006-10-15 01:04:22.000000000 +1000
@@ -72,6 +72,11 @@
 
 #define C_PRIME           (((C - J) * K/256.0) + J)
 #define M_PRIME           (K/256.0 * M)
+
+/* workaround for broken build in 1.6.5.git20061014 */
+# define M_T_PREFERRED 0x08     /* preferred mode within a set */
+# define M_T_DRIVER  0x40       /* Supplied by the driver (EDID, etc) */
+
 /* Established timings from EDID standard */
 static struct
 {
